repository/src/main/java/org/apache/atlas/discovery/EntityLineageService.java

```diff
@@ -357,6 +465,45 @@ private void traverseEdgesOnDemand(Iterable<AtlasEdge> processEdges, boolean isI
         }
     }
 
+    private boolean traverseEdgesOnDemand(Iterator<AtlasEdge> processEdges, boolean isInput, AtlasLineageSizeContext context, AtlasVertex processVertex, int currentDepth, AtomicInteger count) {
+        while (processEdges.hasNext()) {
+            final AtlasEdge processEdge = processEdges.next();
+            AtlasVertex datasetVertex = processEdge.getInVertex();
+
+            boolean isLimitReached = calculateLineageSize(datasetVertex, isInput, context, currentDepth, count);
+            if (isLimitReached)
+                return true;
+        }
+        return false;
+    }
+
+    private void traverseEdgesOnDemand(Iterable<AtlasEdge> processEdges, boolean isInput, int depth, AtlasLineageListContext lineageListContext, AtlasLineageListInfo ret, AtlasVertex processVertex) throws AtlasBaseException {
+        for (AtlasEdge processEdge : processEdges) {
+            AtlasVertex datasetVertex = processEdge.getInVertex();
+
+            if (!vertexMatchesEvaluation(datasetVertex, lineageListContext) || !edgeMatchesEvaluation(processEdge, lineageListContext)) {
+                continue;
+            }
+
+            if (checkForOffset(processVertex, datasetVertex, lineageListContext, ret, isInput)) {
```

Comment:
both above if statements can be combines in one. BTW can we check the offset first and short circuit the other logics before proceeding further.

Reply:
Cannot combine both if statements. Offsets are applicable to filtered entities only. Offset would be applied on a valid filtered dataset so it cannot be checked earlier.